server/appender.go

```diff
@@ -0,0 +1,263 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"math"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/exemplar"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/metadata"
+	"github.com/prometheus/prometheus/model/textparse"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/model/value"
+	"github.com/prometheus/prometheus/storage"
+)
+
+var errNameLabelMandatory = fmt.Errorf("missing metric name (%s label)", labels.MetricName)
+
+type appendErrors struct {
+	numOutOfOrder         int
+	numDuplicates         int
+	numOutOfBounds        int
+	numExemplarOutOfOrder int
+}
+
+// appendFn and scrapecache are copied from prometheus project. Ideally it should've been used from the public API.
+// The public API allows using the ScrapeManager only. We can later use scrape.ScrapeManager but for
+// now we are only using the bits necessary for the project.
+func appendFn(app storage.Appender, logger log.Logger, cache *scrapeCache, b []byte, contentType string, ts time.Time, conf *configuration) error {
+	// textparse.New always returns a valid parser even if there is an error. In that case
+	// it will return the default parser
+	p, err := textparse.New(b, contentType, true)
+	if err != nil {
+		level.Debug(logger).Log(
+			"msg", "Invalid content type on scrape, using prometheus parser as fallback.",
+			"content_type", contentType,
+			"err", err,
+		)
+	}
+
+	var (
+		defTime         = timestamp.FromTime(ts)
+		appErrs         = appendErrors{}
+		sampleLimitErr  error
+		bucketLimitErr  error
+		lset            labels.Labels     // escapes to heap so hoisted out of loop
+		e               exemplar.Exemplar // escapes to heap so hoisted out of loop
+		meta            metadata.Metadata
+		metadataChanged bool
+	)
+
+	// updateMetadata updates the current iteration's metadata object and the
+	// metadataChanged value if we have metadata in the scrape cache AND the
+	// labelset is for a new series or the metadata for this series has just
+	// changed. It returns a boolean based on whether the metadata was updated.
+	updateMetadata := func(lset labels.Labels, isNewSeries bool) bool {
+		if !*conf.EnableMetadataStorage {
+			return false
+		}
+
+		cache.metaMtx.RLock()
+		defer cache.metaMtx.RUnlock()
```

Comment:
Actually, I might be wrong. We are actually modifying the map entry in L71-73.

Reply:
Isn't it reading from the value:
```Go
cache.metadata[lset.Get(labels.MetricName)]
```
The L71-73 just reads it.

Reply:
Ah sorry you are right. I was reviewing this too early in the morning :)